src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.cs

```diff
@@ -0,0 +1,150 @@
+            RuleIdTwoParameterId,
+            s_localizableTitle2,
+            s_localizableMessageDefault2,
+            DiagnosticCategory.Usage,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription2,
+            helpLinkUri: "", // TODO (might require an addition to docs repository?
+            customTags: WellKnownDiagnosticTags.AnalyzerException);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleReplaceOneParameter, RuleReplaceTwoParameter);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(this.OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext startContext)
+        {
+            var compilation = startContext.Compilation;
+
+            var stringSymbol = WellKnownTypes.String(compilation);
+
+            if (stringSymbol is null)
+                return;
+
+            var substring1ParameterMethod = stringSymbol.GetMembers(nameof(string.Substring)).OfType<IMethodSymbol>()
+                .SingleOrDefault(substring => substring.Parameters.Length == 1);
+
+            if (substring1ParameterMethod is null)
+                return;
+
+            var substring2ParameterMethod = stringSymbol.GetMembers(nameof(string.Substring)).OfType<IMethodSymbol>()
+                .SingleOrDefault(substring => substring.Parameters.Length == 2);
+            if (substring2ParameterMethod is null)
+                return;
+
+            var stringBuilderSymbol = WellKnownTypes.StringBuilder(compilation);
+
+            if (stringBuilderSymbol is null)
+                return;
+
+            var sourceAppendMethod = stringBuilderSymbol
+                .GetMembers(nameof(StringBuilder.Append))
+                .OfType<IMethodSymbol>()
+                .Where(append => append.Parameters.Length == 1)
+                .SingleOrDefault(append => append.Parameters[0].Type == stringSymbol);
+
+            if (sourceAppendMethod is null)
+                return;
+            
+            startContext.RegisterOperationAction(
+                context =>
+                {
+                    if (context.Operation is IInvocationOperation invocation)
+                    {
+                        var invokedMethod = invocation.TargetMethod;
+
+                        if (invokedMethod == sourceAppendMethod)
+                        {
+                            var parameters = invokedMethod.Parameters;
+                            if (parameters.Length == 1 && parameters[0].Type.SpecialType == SpecialType.System_String)
```

Comment:
>parameters.Length == 1 && parameters[0].Type.SpecialType == SpecialType.System_String [](start = 32, length = 85)

Seems redundant as you have already checked `invokedMethod == sourceAppendMethod`, and your check is already guaranteed for `sourceAppendMethod`.

Reply:
thanks, fixed.